homeassistant/components/tag/__init__.py

```diff
@@ -0,0 +1,136 @@
+"""The Tag integration."""
+import copy
+import datetime as dt
+import logging
+import typing
+from typing import Optional, cast
+
+import voluptuous as vol
+
+from homeassistant.const import CONF_ID, CONF_NAME
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import collection
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.storage import Store
+from homeassistant.loader import bind_hass
+import homeassistant.util.dt as dt_util
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+DEVICE_ID = "device_id"
+EVENT_TAG_SCANNED = "tag_scanned"
+LAST_SCANNED = "last_scanned"
+STORAGE_KEY = DOMAIN
+STORAGE_VERSION = 1
+TAG_ID = "tag_id"
+TAGS = "tags"
+
+CREATE_FIELDS = {
+    vol.Required(CONF_ID): cv.string,
+    vol.Optional(CONF_NAME): vol.All(str, vol.Length(min=1)),
+    vol.Optional("description"): cv.string,
+    vol.Optional(LAST_SCANNED): cv.datetime,
+}
+
+UPDATE_FIELDS = {
+    vol.Optional(CONF_NAME): vol.All(str, vol.Length(min=1)),
+    vol.Optional("description"): cv.string,
+    vol.Optional(LAST_SCANNED): cv.datetime,
+}
+
+
+class TagIDExistsError(HomeAssistantError):
+    """Raised when an item is not found."""
+
+    def __init__(self, item_id: str):
+        """Initialize tag id exists error."""
+        super().__init__(f"Tag with id: {item_id} already exists.")
+        self.item_id = item_id
+
+
+class TagIDManager(collection.IDManager):
+    """ID manager for tags."""
+
+    def generate_id(self, suggestion: str) -> str:
+        """Generate an ID."""
+        if self.has_id(suggestion):
+            raise TagIDExistsError(suggestion)
+
+        return suggestion
+
+
+class TagStorageCollection(collection.StorageCollection):
+    """Tag collection stored in storage."""
+
+    CREATE_SCHEMA = vol.Schema(CREATE_FIELDS)
+    UPDATE_SCHEMA = vol.Schema(UPDATE_FIELDS)
+
+    async def _process_create_data(self, data: typing.Dict) -> typing.Dict:
+        """Validate the config is valid."""
+        if TAG_ID in data:
+            data[CONF_ID] = data.pop(TAG_ID)
```

Comment:
This is the right approach. We have done a similar thing for Lovelace resources, where we couldn't use "type" because that is already claimed by the websocket message protocol.